rskj-core/src/test/java/org/ethereum/datasource/LevelDbDataSourceFactoryTest.java

```diff
@@ -0,0 +1,89 @@
+package org.ethereum.datasource;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.nio.file.Path;
+import java.util.*;
+
+import static org.hamcrest.Matchers.*;
+
+public class LevelDbDataSourceFactoryTest {
+
+    @Rule
+    public TemporaryFolder databaseDir = new TemporaryFolder();
+
+    private LevelDbDataSourceFactory target;
+
+    @Before
+    public void setUp() {
+        target = new LevelDbDataSourceFactory();
+    }
+
+    @Test
+    public void closedDatasourceDeregisters() {
+        KeyValueDataSource ds1 = target.makeDataSource(databaseDir.getRoot().toPath().resolve("test"));
+        ds1.close();
+        KeyValueDataSource ds2 = target.makeDataSource(databaseDir.getRoot().toPath().resolve("test"));
+
+        Assert.assertThat(ds1, not(equalTo(ds2)));
+    }
+
+    @Test
+    public void retrieveCreatedDataSource() {
+        KeyValueDataSource ds1 = target.makeDataSource(databaseDir.getRoot().toPath().resolve("test"));
+        KeyValueDataSource ds2 = target.makeDataSource(databaseDir.getRoot().toPath().resolve("test"));
+        Assert.assertThat(ds1, equalTo(ds2));
+    }
+
+    @Test
+    public void mergeMultiTrieStoreDBs() {
```

Comment:
I'm not sure if it's the right place, but I haven't found a way to know that every sourceKey is present on the destination